connectors/stash/src/main/java/com/vmware/connectors/stash/StashController.java

```diff
@@ -0,0 +1,444 @@
+
+        final String url = String.format(StashConstants.STASH_PR_ACTION_FORMAT,
+                baseUrl,
+                pullRequest.getProjectKey(),
+                pullRequest.getRepositorySlug(),
+                pullRequest.getPullRequestId(),
+                stashAction.getAction(),
+                version);
+
+        // This header is added for skipping the CSRF check. Otherwise, Atlassian denies the pull request action.
+        headers.add(ATLASSIAN_TOKEN, "no-check");
+
+        final ListenableFuture<ResponseEntity<String>> response = this.rest.exchange(
+                url,
+                HttpMethod.POST,
+                new HttpEntity<>(headers),
+                String.class);
+        return Async.toSingle(response);
+    }
+
+    private HttpHeaders buildHeaders(final String authHeader) {
+        final HttpHeaders headers = new HttpHeaders();
+        headers.add(AUTHORIZATION, authHeader);
+        return headers;
+    }
+
+    private String getVersion(final HttpHeaders headers,
+                              final String baseUrl,
+                              final StashPullRequest pullRequest) {
+
+        final ListenableFuture<ResponseEntity<JsonDocument>> response = getPullRequestInfo(headers, pullRequest, baseUrl);
+        final JsonDocument[] jsonDocuments = new JsonDocument[1];
+        Async.toSingle(response)
+                .subscribe(entity -> {
+                    jsonDocuments[0] = entity.getBody();
+                });
+
+        Assert.isTrue(Objects.nonNull(jsonDocuments[0]), "JsonDocument should not be null.");
+        return Integer.toString(jsonDocuments[0].read("$.version"));
+    }
+
+    private Observable<Card> getCardsForStashPR(final HttpHeaders headers,
+                                                final StashPullRequest pullRequest,
+                                                final String baseUrl,
+                                                final String routingPrefix) {
+        logger.debug("Requesting pull request info from stash base url: {} and pull request info: {}", baseUrl, pullRequest);
+
+        final ListenableFuture<ResponseEntity<JsonDocument>> stashResponse = getPullRequestInfo(headers, pullRequest, baseUrl);
+        final List<String> comments = getComments(baseUrl, headers, pullRequest);
+
+        return Async.toSingle(stashResponse)
+                .toObservable()
+                .onErrorResumeNext(ObservableUtil::skip404)
+                .map(entity -> convertResponseIntoCard(entity, pullRequest, routingPrefix, comments));
+    }
+
+    private ListenableFuture<ResponseEntity<JsonDocument>> getPullRequestInfo(final HttpHeaders headers,
+                                                                              final StashPullRequest pullRequest,
+                                                                              final String baseUrl) {
+        final String url = String.format(StashConstants.STASH_PR_URL_FORMAT,
```

Comment:
Another String.format to convert to UriComponentsBuilder.

Reply:
I have used ```UriComponentsBuilder``` instead of ```String.format``` in all the places..